src/backend/looped/vehicle/vehicle_ammo_special_type.cpp

```diff
@@ -0,0 +1,248 @@
+			m_mg_range               = weapon_info->m_weapon_range;
+		}
+
+		void apply_mg(const vehicle_ammo_setting& g_vehicle_ammo_settings)
+		{
+			// apply ammo type changes to the current weapon
+			eDamageType damage_type     = eDamageType::None;
+			eExplosionTag explosion_tag = g_vehicle_ammo_settings.explosion_tag;
+			eAmmoSpecialType ammo_type  = eAmmoSpecialType::None;
+
+			if (explosion_tag == eExplosionTag::DONTCARE)
+			{
+				damage_type = m_mg_damage_type;
+				ammo_type   = g_vehicle_ammo_settings.type;
+			}
+			else
+			{
+				damage_type = eDamageType::Explosive;
+				ammo_type   = m_mg_ammo_type;
+			}
+
+			m_mg_weapon_info->m_damage_type = damage_type;
+
+			CWeaponInfo::sExplosion explosion;
+			explosion.m_default                                = explosion_tag;
+			explosion.m_hit_bike                               = explosion_tag;
+			explosion.m_hit_boat                               = explosion_tag;
+			explosion.m_hit_car                                = explosion_tag;
+			explosion.m_hit_plane                              = explosion_tag;
+			explosion.m_hit_truck                              = explosion_tag;
+			m_mg_weapon_info->m_explosion                      = explosion;
+			m_mg_weapon_info->m_ammo_info->m_ammo_special_type = ammo_type;
+			m_mg_weapon_info->m_speed                          = g_vehicle_ammo_settings.speed;
+			m_mg_weapon_info->m_time_between_shots             = g_vehicle_ammo_settings.time_between_shots;
+			m_mg_weapon_info->m_alternate_wait_time            = g_vehicle_ammo_settings.alternate_wait_time;
+			m_mg_weapon_info->m_weapon_range                   = g_vehicle_ammo_settings.weapon_range;
+		}
+
+		void restore_rocket()
+		{
+			if (m_rocket_weapon_info != nullptr)
+			{
+				m_rocket_weapon_info->m_damage_type         = m_rocket_damage_type;
+				m_rocket_weapon_info->m_time_between_shots  = m_rocket_time_between_shots;
+				m_rocket_weapon_info->m_alternate_wait_time = m_rocket_alternate_wait_time;
+				m_rocket_weapon_info->m_lock_on_range       = m_rocket_lock_on_range;
+				m_rocket_weapon_info->m_weapon_range        = m_rocket_range;
+				m_rocket_weapon_info->m_reload_time_mp      = m_rocket_reload_time_mp;
+				m_rocket_weapon_info->m_reload_time_sp      = m_rocket_reload_time_sp;
+				weapon_flags(m_rocket_weapon_info)          = m_rocket_weapon_flags;
+
+				CAmmoRocketInfo* rocket_info        = (CAmmoRocketInfo*)m_rocket_weapon_info->m_ammo_info;
+				rocket_info->m_explosion            = m_rocket_explosion;
+				rocket_info->m_lifetime             = m_rocket_lifetime;
+				rocket_info->m_launch_speed         = m_rocket_launch_speed;
+				rocket_info->m_time_before_homing   = m_rocket_time_before_homing;
+				rocket_info->m_homing_rocket_params = m_rocket_homing_params;
+				m_rocket_weapon_info                = nullptr;
+			}
+		}
+
+		void backup_rocket(CWeaponInfo* weapon_info)
+		{
+			m_rocket_weapon_info         = weapon_info;
```

Comment:
Why are you checking if m_rocket_weapon_info is not a null pointer before running the code in restore_rocket but not in backup_rocket? Is it impossible for m_rocket_weapon_info to be in a dirty state for backup_rocket?

Reply:
It is impossible because 
1. weapon_info is guaranteed != null
2. if m_mg_weapon_info is null, then [m_mg_weapon_info != weapon_info](https://github.com/George0828Zhang/YimMenu/blob/f1c2dfdd6836e4ac8695d9dda8c9fe516c6829d6/src/backend/looped/vehicle/vehicle_ammo_special_type.cpp#L49), meaning unarmed -> equipped weapon
3. backup_mg(weapon_info) would've been called, setting m_mg_weapon_info to valid value.
The null check in restore() is required, otherwise error arise when you toggle off without ever changing weapons.